api/service/stagedsync/stage.go

```diff
@@ -0,0 +1,180 @@
+	DisabledDescription string
+	// Disabled defines if the stage is disabled. It sets up when the stage is build by its `StageBuilder`.
+	Disabled bool
+}
+
+var ErrStopped = errors.New("stopped")
+var ErrUnwind = errors.New("unwound")
+
+// StageState is the state of the stage.
+type StageState struct {
+	state       *StagedSync
+	ID          SyncStageID
+	BlockNumber uint64 // BlockNumber is the current block number of the stage at the beginning of the state execution.
+}
+
+func (s *StageState) LogPrefix() string { return s.state.LogPrefix() }
+
+func (s *StageState) CurrentStageProgress(db kv.Getter) (uint64, error) {
+	return GetStageProgress(db, s.ID, s.state.isBeacon)
+}
+
+func (s *StageState) StageProgress(db kv.Getter, id SyncStageID) (uint64, error) {
+	return GetStageProgress(db, id, s.state.isBeacon)
+}
+
+// Update updates the stage state (current block number) in the database. Can be called multiple times during stage execution.
+func (s *StageState) Update(db kv.Putter, newBlockNum uint64) error {
+	// if m, ok := syncMetrics[s.ID]; ok {
+	// 	m.Set(newBlockNum)
+	// }
+	return SaveStageProgress(db, s.ID, s.state.isBeacon, newBlockNum)
+}
+func (s *StageState) UpdatePrune(db kv.Putter, blockNum uint64) error {
+	return SaveStagePruneProgress(db, s.ID, s.state.isBeacon, blockNum)
+}
+
+// Unwinder allows the stage to cause an unwind.
+type Unwinder interface {
+	// UnwindTo begins staged sync unwind to the specified block.
+	UnwindTo(unwindPoint uint64, invalidBlock common.Hash)
+}
+
+// UnwindState contains the information about unwind.
+type UnwindState struct {
+	ID SyncStageID
+	// UnwindPoint is the block to unwind to.
+	UnwindPoint        uint64
+	CurrentBlockNumber uint64
+	// If unwind is caused by a bad block, this hash is not empty
+	InvalidBlock common.Hash
+	state        *StagedSync
+}
+
+func (u *UnwindState) LogPrefix() string { return u.state.LogPrefix() }
+
+// Done updates the DB state of the stage.
+func (u *UnwindState) Done(db kv.Putter) error {
+	return SaveStageProgress(db, u.ID, u.state.isBeacon, u.UnwindPoint)
+}
+
+type PruneState struct {
+	ID              SyncStageID
+	ForwardProgress uint64 // progress of stage forward move
+	PruneProgress   uint64 // progress of stage prune move. after sync cycle it become equal to ForwardProgress by Done() method
+	state           *StagedSync
+}
+
+func (s *PruneState) LogPrefix() string { return s.state.LogPrefix() + " Prune" }
+func (s *PruneState) Done(db kv.Putter) error {
+	return SaveStagePruneProgress(db, s.ID, s.state.isBeacon, s.ForwardProgress)
+}
+func (s *PruneState) DoneAt(db kv.Putter, blockNum uint64) error {
+	return SaveStagePruneProgress(db, s.ID, s.state.isBeacon, blockNum)
+}
+
+// PruneTable has `limit` parameter to avoid too large data deletes per one sync cycle - better delete by small portions to reduce db.FreeList size
+func PruneTable(tx kv.RwTx, table string, pruneTo uint64, ctx context.Context, limit int) error {
```

Comment:
It will be useful in next version which we need to prune some tables. Actually I didn't remove it to be a part of standard. Very likely in this version also will be used